src/java_bytecode/java_types.h

```diff
@@ -76,4 +78,124 @@ bool is_java_array_tag(const irep_idt& tag);
 
 bool is_valid_java_array(const struct_typet &type);
 
+
+/// class to hold a Java generic type
+/// upper bound can specify type requirements
+class java_generic_typet:public reference_typet
```

Comment:
I think I get it - this is type that is a generic parameter (either with a bound or it is a `java_inst_generic_typet` if we know the actual type). Consider renaming to `java_generic_parameter_typet` since a generic type in Java normally means a type that has generic parameters rather than the parameters themselves.